addons/stock_account/models/product.py

```diff
@@ -310,11 +310,23 @@ def _run_fifo(self, quantity, company):
 
         # Find back incoming stock valuation layers (called candidates here) to value `quantity`.
         qty_to_take_on_candidates = quantity
-        candidates = self.env['stock.valuation.layer'].sudo().search([
-            ('product_id', '=', self.id),
-            ('remaining_qty', '>', 0),
-            ('company_id', '=', company.id),
-        ])
+        location = self.env.context.get('location')
+        condidates = self.env['stock.valuation.layer']
+        if location:
+            candidates = self.env['stock.valuation.layer'].sudo().search([
+                ('product_id', '=', self.id),
+                ('remaining_qty', '>', 0),
+                ('company_id', '=', company.id),
+                ('stock_move_id', 'in', self.env['stock.move'].sudo().search([
```

Comment:
this query can be horrible in large database

You can replace by :
```python
self.env['stock.valuation.layer'].sudo().search([
                ('product_id', '=', self.id),
                ('remaining_qty', '>', 0),
                ('company_id', '=', company.id),
                ('stock_move_id.location_dest_id', 'in',location.id}]}
```

Reply:
Thanks